aquadoggo/src/graphql/replication/query.rs

```diff
@@ -0,0 +1,421 @@
+        runner: TestDatabaseRunner,
+    ) {
+        runner.with_db_teardown(|db: TestDatabase| async move {
+            let replication_root = ReplicationRoot::default();
+            let schema = Schema::build(replication_root, EmptyMutation, EmptySubscription)
+                .data(db.store)
+                .finish();
+
+            // The test runner creates a test entry for us. The entry hash is automatically the
+            // document id as it contains the CREATE operation
+            let entry_hash_str = db.test_data.documents.first().unwrap().as_str();
+
+            // Construct the query
+            let gql_query = format!(
+                r#"
+                    query {{
+                        entryByHash(hash: "{}") {{
+                            entry
+                            operation
+                        }}
+                    }}
+                "#,
+                entry_hash_str
+            );
+
+            // Make the query
+            let result = schema.execute(Request::new(gql_query.clone())).await;
+
+            // Check that query was successful
+            assert!(
+                result.is_ok(),
+                "Query: {} \nResult: {:?}",
+                gql_query,
+                result
+            );
+        });
+    }
+
+    #[rstest]
+    fn entry_by_hash_not_found(
+        #[from(test_db)] runner: TestDatabaseRunner,
+        #[from(random_hash)] random_hash: Hash,
+    ) {
+        runner.with_db_teardown(|db: TestDatabase| async move {
+            let replication_root = ReplicationRoot::default();
+            let schema = Schema::build(replication_root, EmptyMutation, EmptySubscription)
+                .data(db.store)
+                .finish();
+
+            let gql_query = format!(
+                r#"query {{ entryByHash(hash: "{}") {{ entry operation }} }}"#,
+                random_hash.as_str()
+            );
+
+            // Make sure that query returns an error as entry was not found
+            let result = schema.execute(Request::new(gql_query.clone())).await;
+            assert!(result.is_err(), "{:?}", result);
+        });
+    }
+
+    #[rstest]
+    fn entry_by_log_id_and_seq_num(
+        #[from(test_db)]
+        #[with(1, 1, 1)]
+        runner: TestDatabaseRunner,
+    ) {
+        runner.with_db_teardown(|db: TestDatabase| async move {
+            let replication_root = ReplicationRoot::default();
+            let schema = Schema::build(replication_root, EmptyMutation, EmptySubscription)
+                .data(db.store)
+                .finish();
+
+            // The test runner creates a test entry for us, we can retreive the public key from the
+            // author
+            let public_key: Author = db
+                .test_data
+                .key_pairs
+                .first()
+                .unwrap()
+                .public_key()
+                .to_owned()
+                .try_into()
+                .unwrap();
```

Comment:
This is so ugly... :rofl: I think we need some more conversion traits for `Author`.